File: apng/src/main/java/com/github/penfeizhou/animation/apng/decode/StillFrame.java
package com.github.penfeizhou.animation.apng.decode;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Paint;

import androidx.annotation.NonNull;

import com.github.penfeizhou.animation.apng.io.APNGReader;
import com.github.penfeizhou.animation.decode.Frame;
import com.github.penfeizhou.animation.io.Writer;

import java.io.IOException;

/**
 * @Description: APNG4Android
 * @Author: pengfei.zhou
 * @CreateDate: 2019-05-13
 */
public class StillFrame extends Frame<Writer> {
    @NonNull
    private final APNGReader reader;

    public StillFrame(@NonNull APNGReader reader) {
        this.reader = reader;
    }

    @Override
    public Bitmap draw(Canvas canvas, Paint paint, int sampleSize, Bitmap reusedBitmap, Writer writer) {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = false;
        options.inSampleSize = sampleSize;
        options.inMutable = true;
        options.inBitmap = reusedBitmap;
        Bitmap bitmap = null;
        try {
            reader.reset();

            try {
                bitmap = BitmapFactory.decodeStream(reader.toInputStream(), null, options);
            } catch (IllegalArgumentException e) {
                // Problem decoding into existing bitmap when on Android 4.2.2 & 4.3
                BitmapFactory.Options optionsFixed = new BitmapFactory.Options();
                optionsFixed.inJustDecodeBounds = false;
                optionsFixed.inSampleSize = sampleSize;
                optionsFixed.inMutable = true;
                bitmap = BitmapFactory.decodeStream(reader.toInputStream(), null, optionsFixed);
            }
            assert bitmap != null;
            paint.setXfermode(null);
            canvas.drawBitmap(bitmap, 0, 0, paint);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return bitmap;
    }
}

File: apng/src/main/java/com/github/penfeizhou/animation/apng/decode/StillFrame.kt
package com.github.penfeizhou.animation.apng.decode

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import com.github.penfeizhou.animation.decode.Frame
import com.github.penfeizhou.animation.io.FilterReader
import com.github.penfeizhou.animation.io.Writer
import java.io.IOException

internal class StillFrame(
    private val reader: FilterReader,
    width: Int,
    height: Int
) : Frame(x = 0, y = 0, width = width, height = height, duration = 0) {

    override fun draw(
        canvas: Canvas,
        paint: Paint,
        sampleSize: Int,
        reusedBitmap: Bitmap,
        writer: Writer
    ): Bitmap? {
        val options = createBitmapFactoryOptions(sampleSize, reusedBitmap)
        var bitmap: Bitmap? = null
        try {
            reader.reset()
            bitmap = try {
                BitmapFactory.decodeStream(reader.toInputStream(), null, options)
            } catch (e: IllegalArgumentException) {
                // Problem decoding into existing bitmap when on Android 4.2.2 & 4.3
                val optionsFixed = createBitmapFactoryOptions(sampleSize, null)
                BitmapFactory.decodeStream(reader.toInputStream(), null, optionsFixed)
            }
            assert(bitmap != null)
            paint.xfermode = null
            canvas.drawBitmap(bitmap!!, 0f, 0f, paint)
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return bitmap
    }

    private fun createBitmapFactoryOptions(
        sampleSize: Int,
        reusedBitmap: Bitmap?
    ): BitmapFactory.Options = BitmapFactory.Options().apply {
        inJustDecodeBounds = false
        inSampleSize = sampleSize
        inMutable = true
        inBitmap = reusedBitmap
    }
}
